plutus-tx/src/PlutusTx/BuiltinList.hs

```diff
@@ -5,81 +5,489 @@
+      let BI.BuiltinPair (xs'', xs''') = _splitAt (n `B.subtractInteger` 1) xs'
+      BI.BuiltinPair (x <| xs'', xs''')
+{-# INLINABLE _splitAt #-}
+
+-- | Plutus Tx version of 'Data.List.nub' for 'BuiltinList'.
+nub :: forall a. (Eq a, MkNil a) => BuiltinList a -> BuiltinList a
+nub = nubBy (==)
+{-# INLINABLE nub #-}
+
+-- | Plutus Tx version of 'Data.List.elemBy' for 'BuiltinList'.
+elemBy :: forall a. (a -> a -> Bool) -> a -> BuiltinList a -> Bool
+elemBy eq y = go
+  where
+    go :: BuiltinList a -> Bool
+    go = caseList' False ( \x xs -> if eq x y then True else go xs )
+{-# INLINABLE elemBy #-}
+
+-- | Plutus Tx version of 'Data.List.nubBy' for 'BuiltinList'.
+nubBy :: forall a. (MkNil a) => (a -> a -> Bool) -> BuiltinList a -> BuiltinList a
+nubBy eq = flip go empty
+  where
+    go :: BuiltinList a -> BuiltinList a -> BuiltinList a
+    go l xs = caseList' empty
+      ( \y ys ->
+          if elemBy eq y xs
+          then go ys xs
+          else y <| go ys (y <| xs)
+      ) l
+{-# INLINABLE nubBy #-}
+
+-- | Plutus Tx version of 'Data.List.zipWith' for 'BuiltinList'.
+zipWith
+  :: forall a b c. (MkNil c)
+  => (a -> b -> c)
+  -> BuiltinList a
+  -> BuiltinList b
+  -> BuiltinList c
+zipWith f = go
+  where
+    go :: BuiltinList a -> BuiltinList b -> BuiltinList c
+    go xs ys =
+      caseList' empty
+        ( \x xs' ->
+            caseList' empty
+              ( \y ys' -> f x y <| go xs' ys'
+              ) ys
+        ) xs
+{-# INLINABLE zipWith #-}
+
+-- | Plutus Tx version of 'Data.List.dropWhile' for 'BuiltinList'.
+dropWhile :: forall a. (MkNil a) => (a -> Bool) -> BuiltinList a -> BuiltinList a
+dropWhile p = go
+  where
+    go :: BuiltinList a -> BuiltinList a
+    go xs = caseList' empty ( \x xs' -> if p x then go xs' else xs ) xs
+{-# INLINABLE dropWhile #-}
+
+-- | Plutus Tx version of 'Data.List.replicate' for 'BuiltinList'.
+replicate :: forall a. (MkNil a) => Integer -> a -> BuiltinList a
+replicate n0 x = go n0
+  where
+    go :: Integer -> BuiltinList a
+    go n
+      | n `B.lessThanEqualsInteger` 0 = empty
+      | otherwise = x <| go (B.subtractInteger n 1)
+{-# INLINABLE replicate #-}
+
+-- | Plutus Tx version of 'Data.List.partition' for 'BuiltinList'.
+_partition
+  :: forall a. (MkNil a)
+  => (a -> Bool)
+  -> BuiltinList a
+  -> BuiltinPair (BuiltinList a) (BuiltinList a)
+_partition p = BI.BuiltinPair . foldr select (empty, empty)
+  where
+    select :: a -> (BuiltinList a, BuiltinList a) -> (BuiltinList a, BuiltinList a)
+    select x ~(ts, fs) = if p x then (x <| ts, fs) else (ts, x <| fs)
+{-# INLINABLE _partition #-}
+
+-- | Plutus Tx version of 'Data.List.sort' for 'BuiltinList'.
+_sort :: (MkNil a, Ord a) => BuiltinList a -> BuiltinList a
+_sort = _sortBy compare
+{-# INLINABLE _sort #-}
+
+-- | Plutus Tx version of 'Data.List.sortBy' for 'BuiltinList'.
+_sortBy :: MkNil a => (a -> a -> Ordering) -> BuiltinList a -> BuiltinList a
+_sortBy cmp = mergeAll . sequences
```

Comment:
It looks like for this one you can perform the same trick of reusing the empty list as with `dropWhile` to get rid of the `MkNil` constraint.